packages/vhd-lib/src/createVhdDirectoryFromStream.js

```diff
@@ -45,72 +45,69 @@ async function* parse(stream) {
   const blockSize = header.blockSize
   const blockAndBitmapSize = blockBitmapSize + blockSize
 
-  const batOffset = header.tableOffset
-  const batSize = Math.max(1, Math.ceil((header.maxTableEntries * 4) / SECTOR_SIZE)) * SECTOR_SIZE
-  /**
-   * the norm allows the BAT to be after some blocks or parent locator
-   * we do not handle this case for now since we need the BAT to order the blocks/parent locator
-   *
-   * also there can be some free space between header and the start of BAT
-   */
-
-  const bat = await read(batOffset, batSize)
-
-  /**
-   * the norm allows blocks and parent locators  to be intervined
-   *
-   * we build a sorted index since we read the stream sequentially and
-   * we need to know what is the the next element to read and its size
-   * (parent locator size can vary)
-   */
-
   const index = []
-  for (let blockCounter = 0; blockCounter < header.maxTableEntries; blockCounter++) {
-    const batEntrySector = bat.readUInt32BE(blockCounter * 4)
-    // unallocated block, no need to export it
-    if (batEntrySector === BLOCK_UNUSED) {
-      continue
-    }
-    const batEntryBytes = batEntrySector * SECTOR_SIZE
-    // ensure the block is not before the bat
-    assert.ok(batEntryBytes >= batOffset + batSize)
-    index.push({
-      type: 'block',
-      id: blockCounter,
-      offset: batEntryBytes,
-      size: blockAndBitmapSize,
-    })
-  }
 
   for (const parentLocatorId in header.parentLocatorEntry) {
     const parentLocatorEntry = header.parentLocatorEntry[parentLocatorId]
     // empty parent locator entry, does not exist in the content
     if (parentLocatorEntry.platformDataSpace === 0) {
       continue
     }
-    assert.ok(parentLocatorEntry.platformDataOffset * SECTOR_SIZE >= batOffset + batSize)
-
     index.push({
+      ...parentLocatorEntry,
       type: 'parentLocator',
-      offset: parentLocatorEntry.platformDataOffset * SECTOR_SIZE,
-      size: parentLocatorEntry.platformDataSpace * SECTOR_SIZE,
+      offset: parentLocatorEntry.platformDataOffset,
+      size: parentLocatorEntry.platformDataLength,
       id: parentLocatorId,
     })
   }
 
-  index.sort((a, b) => a.offset - b.offset)
+  const batOffset = header.tableOffset
+  const batSize = Math.max(1, Math.ceil((header.maxTableEntries * 4) / SECTOR_SIZE)) * SECTOR_SIZE
 
-  for (const item of index) {
+  index.push({
+    type: 'bat',
+    offset: batOffset,
+    size: batSize,
+  })
+
+  // sometimes some parent locator are before the BAT
+  index.sort((a, b) => a.offset - b.offset)
+  let item
```

Comment:
Discussion: I love languages in which you can declare a variable directly in the condition of a `if` or a `while`.

Reply:
Are you sure that you can declare a variable in `while` ?

Reply:
No, it's not possible in ECMAScript, I was just saying that I like languages that provides this :slightly_smiling_face: